File: src/backend/services/web/risk/resources/event.py
# -*- coding: utf-8 -*-
"""
TencentBlueKing is pleased to support the open source community by making
蓝鲸智云 - 审计中心 (BlueKing - Audit Center) available.
Copyright (C) 2023 THL A29 Limited,
a Tencent company. All rights reserved.
Licensed under the MIT License (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at http://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
either express or implied. See the License for the
specific language governing permissions and limitations under the License.
We undertake not to change the open source license (MIT license) applicable
to the current version of the project delivered to anyone in the future.
"""

import datetime
import json
import uuid
from typing import List

from django.shortcuts import get_object_or_404
from django.utils import timezone
from django.utils.translation import gettext, gettext_lazy
from rest_framework import serializers

from apps.audit.resources import AuditMixinResource
from apps.meta.models import GlobalMetaConfig
from core.utils.tools import get_app_info
from services.web.risk.constants import (
    ENABLE_EVENTS_MOCK_KEY,
    EVENTS_MOCK_DATA_KEY,
    EventMappingFields,
    RiskStatus,
)
from services.web.risk.handlers import EventHandler
from services.web.risk.handlers.risk import RiskHandler
from services.web.risk.models import Risk
from services.web.risk.permissions import GenerateStrategyRiskPermission
from services.web.risk.serializers import (
    CreateEventAPIResponseSerializer,
    CreateEventAPISerializer,
    ListEventRequestSerializer,
    ListEventResponseSerializer,
)
from services.web.risk.tasks import manual_add_event
from services.web.strategy_v2.constants import StrategyType
from services.web.strategy_v2.models import Strategy


class EventMeta(AuditMixinResource):
    tags = ["Event"]


class CreateEvent(EventMeta):
    name = gettext_lazy("创建事件")
    RequestSerializer = CreateEventAPISerializer
    ResponseSerializer = CreateEventAPIResponseSerializer
    bind_request = True

    def _create_events(self, events: List[dict], source: str):
        event_ids = []
        for event in events:
            event["event_id"] = "{}{}".format(event.get("strategy_id"), event["raw_event_id"] or uuid.uuid1().hex)
            event.setdefault("event_source", source)
            event_ids.append(event["event_id"])
        return event_ids

    def perform_request(self, validated_request_data: dict):
        gen_risk = validated_request_data.get("gen_risk", False)
        events = validated_request_data["events"]
        if not gen_risk:
            bound_risk = self._validate_existing_risk(validated_request_data.get("risk_id"), events)
            self._sync_raw_event_id_with_risk(events, bound_risk)
            self._ensure_rule_strategy(bound_risk.strategy_id)
        else:
            for event in events:
                self._ensure_rule_strategy(event["strategy_id"])
        for event in events:
            self._apply_basic_field_mapping(event)
        req = validated_request_data.get("_request")
        if req and getattr(req, "user", None) and getattr(req.user, "is_authenticated", False):
            GenerateStrategyRiskPermission(req).ensure_allowed(events)
            source = req.user.username
        else:
            app = get_app_info()
            source = app.bk_app_code
        event_ids = self._create_events(events, source)
        manual_add_event(events)
        risk_ids = []
        eligible_strategy_ids = RiskHandler.fetch_eligible_strategy_ids()  # 更新 eligible_strategy_ids
        for event in events:
            risk_id = RiskHandler().generate_risk(event, eligible_strategy_ids, manual=True)
            if risk_id:
                risk_ids.append(risk_id)
        return {"event_ids": event_ids, "risk_ids": risk_ids}

    def _validate_existing_risk(self, risk_id: str, events: List[dict]) -> Risk:
        if not risk_id:
            raise serializers.ValidationError(gettext("风险ID不能为空"))

        risk = (
            Risk.objects.filter(risk_id=str(risk_id))
            .only("status", "event_time", "event_end_time", "strategy_id")
            .first()
        )
        if not risk:
            raise serializers.ValidationError(gettext("风险不存在或不可用"))

        tz = timezone.get_current_timezone()
        for event in events:
            strategy_id = event["strategy_id"]
            if strategy_id and int(strategy_id) != risk.strategy_id:
                raise serializers.ValidationError(gettext("事件所属策略与风险不一致"))

            if risk.status == RiskStatus.CLOSED:
                raise serializers.ValidationError(gettext("风险单已关闭，无法添加事件"))

            event_dt = datetime.datetime.fromtimestamp(event["event_time"] / 1000, tz=tz)
            if not risk.event_time <= event_dt:
                raise serializers.ValidationError(gettext("事件时间不在风险有效区间内"))
        return risk

    def _sync_raw_event_id_with_risk(self, events: List[dict], risk: Risk):
        for event in events:
            event["raw_event_id"] = risk.raw_event_id

    def _ensure_rule_strategy(self, strategy_id: int):
        strategy = Strategy.objects.filter(strategy_id=strategy_id).only("strategy_type").first()
        if not strategy:
            raise serializers.ValidationError(gettext("策略不存在"))
        if strategy.strategy_type != StrategyType.RULE:
            raise serializers.ValidationError(gettext("仅支持规则审计策略创建事件"))

    def _apply_basic_field_mapping(self, event: dict) -> None:
        """
        手工创建事件时，按策略基础字段映射补充事件字段
        """

        strategy = (
            Strategy.objects.filter(strategy_id=event.get("strategy_id")).only("event_basic_field_configs").first()
        )
        if not strategy:
            return

        basic_configs = strategy.event_basic_field_configs or []
        if not basic_configs:
            return

        raw_event_data = event.get("event_data") or {}
        if isinstance(raw_event_data, str):
            try:
                event_data = json.loads(raw_event_data)
            except json.JSONDecodeError:
                return
        elif isinstance(raw_event_data, dict):
            event_data = raw_event_data
        else:
            return

        for field in basic_configs:
            field_name = field.get("field_name")
            map_config = field.get("map_config") or {}
            if not field_name or not map_config:
                continue

            if field_name in [
                EventMappingFields.RAW_EVENT_ID.field_name,
                EventMappingFields.EVENT_TIME.field_name,
                EventMappingFields.STRATEGY_ID.field_name,
            ]:
                continue

            target_value = map_config.get("target_value")
            source_field = map_config.get("source_field")
            if target_value is not None:
                mapped_value = target_value
            elif source_field:
                mapped_value = event_data.get(source_field)
            else:
                continue

            if field_name == EventMappingFields.EVENT_TYPE.field_name:
                event[field_name] = (
                    mapped_value
                    if isinstance(mapped_value, str)
                    else ",".join(map(str, mapped_value))
                    if isinstance(mapped_value, (list, tuple, set))
                    else str(mapped_value)
                )
                continue

            if field_name == EventMappingFields.OPERATOR.field_name:
                event[field_name] = (
                    mapped_value
                    if isinstance(mapped_value, str)
                    else ",".join(map(str, mapped_value))
                    if isinstance(mapped_value, (list, tuple, set))
                    else str(mapped_value)
                )
                continue

            event[field_name] = mapped_value


class ListEvent(EventMeta):
    name = gettext_lazy("获取事件列表")
    RequestSerializer = ListEventRequestSerializer
    ResponseSerializer = ListEventResponseSerializer

    def perform_request(self, validated_request_data):
        risk = get_object_or_404(Risk, risk_id=validated_request_data.pop("risk_id"))
        _ = validated_request_data.pop("_request", None)
        validated_request_data.update({"raw_event_id": risk.raw_event_id, "strategy_id": str(risk.strategy_id)})
        if GlobalMetaConfig.get(ENABLE_EVENTS_MOCK_KEY, default=False) and GlobalMetaConfig.get(
            EVENTS_MOCK_DATA_KEY, default=[]
        ):
            return {
                "page": 1,
                "num_pages": 50,
                "total": 1,
                "results": GlobalMetaConfig.get(EVENTS_MOCK_DATA_KEY),
            }
        return EventHandler.search_event(**validated_request_data)

File: src/frontend/src/domain/model/strategy/event.ts
/*
  TencentBlueKing is pleased to support the open source community by making
  蓝鲸智云 - 审计中心 (BlueKing - Audit Center) available.
  Copyright (C) 2023 THL A29 Limited,
  a Tencent company. All rights reserved.
  Licensed under the MIT License (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at http://opensource.org/licenses/MIT
  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on
  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
  either express or implied. See the License for the
  specific language governing permissions and limitations under the License.
  We undertake not to change the open source license (MIT license) applicable
  to the current version of the project delivered to anyone in the future.
*/
const STATUS_ABNORMAL = 'ABNORMAL'; // 未恢复
const STATUS_CLOSED = 'CLOSED'; // 已关闭
const STATUS_RECOVERED = 'RECOVERED'; // 已恢复
const STATUS_UNKNOWN = 'UNKNOWN'; // 未知

export default class Event {
  static statusMap: Record<string, string> = {
    [STATUS_ABNORMAL]: 'abnormal',
    [STATUS_CLOSED]: 'close',
    [STATUS_RECOVERED]: 'recovered',
    [STATUS_UNKNOWN]: 'unknown',
  };

  id: string;
  alert_name: string;
  status: string;
  labels: Array<string>;
  description: string;
  severity: number;
  metric: Array<string>;
  bk_biz_id: number;
  ip: string;
  bk_cloud_id: string;
  bk_service_instance_id: string;
  bk_topo_node: string;
  assignee: Array<string>;
  appointee: string;
  is_ack: string;
  is_shielded: boolean;
  shield_left_time: string;
  shield_id: string;
  is_handled: boolean;
  strategy_id: number;
  create_time: string;
  update_time: string;
  begin_time: string;
  end_time: string;
  latest_time: string;
  first_anomaly_time: string;
  target_type: string;
  target: string;
  category: string;
  tags: Array<
    {
      key: string;
      value: string;
    }
  >;
  category_display: string;
  duration: string;
  ack_duration: string;
  data_type: string;
  converge_id: string;
  event_id: string;
  plugin_id: string;
  stage_display: string;
  dimensions: Array<{
    display_value: string;
    display_key: string;
    value: string;
    key: string;
  }>;
  seq_id: number;
  dedupe_md5: string;
  dedupe_keys: [string];
  dimension_message: string;
  metric_display: Array<
    {
      id: string;
      name: string
    }
  >;
  target_key: string;
  ack_operator: string;
  shield_operator: Array<string>;
  strategy_name: string;
  bk_biz_name: string;
  status_name: string;


  constructor(payload = {} as Event) {
    this.id = payload.id;
    this.alert_name = payload.alert_name;
    this.status = payload.status;
    this.labels = payload.labels || [];
    this.description = payload.description;
    this.severity = payload.severity;
    this.metric = payload.metric;
    this.bk_biz_id = payload.bk_biz_id;
    this.ip = payload.ip;
    this.bk_cloud_id = payload.bk_cloud_id;
    this.bk_service_instance_id = payload.bk_service_instance_id;
    this.bk_topo_node = payload.bk_topo_node;
    this.assignee = payload.assignee;
    this.appointee = payload.appointee;
    this.is_ack = payload.is_ack;
    this.is_shielded = payload.is_shielded;
    this.shield_left_time = payload.shield_left_time;
    this.shield_id = payload.shield_id;
    this.is_handled = payload.is_handled;
    this.strategy_id = payload.strategy_id;
    this.create_time = payload.create_time;
    this.update_time = payload.update_time;
    this.begin_time = payload.begin_time;
    this.end_time = payload.end_time;
    this.latest_time = payload.latest_time;
    this.first_anomaly_time = payload.first_anomaly_time;
    this.target_type = payload.target_type;
    this.target = payload.target;
    this.category = payload.category;
    this.tags = payload.tags;
    this.category_display = payload.category_display;
    this.duration = payload.duration;
    this.ack_duration = payload.ack_duration;
    this.data_type = payload.data_type;
    this.converge_id = payload.converge_id;
    this.event_id = payload.event_id;
    this.plugin_id = payload.plugin_id;
    this.stage_display = payload.stage_display;
    this.dimensions = payload.dimensions;
    this.seq_id = payload.seq_id;
    this.dedupe_md5 = payload.dedupe_md5;
    this.dedupe_keys = payload.dedupe_keys;
    this.dimension_message = payload.dimension_message;
    this.metric_display = payload.metric_display || [];
    this.target_key = payload.target_key;
    this.ack_operator = payload.ack_operator;
    this.shield_operator = payload.shield_operator;
    this.strategy_name = payload.strategy_name;
    this.bk_biz_name = payload.bk_biz_name;
    this.status_name = payload.status_name;
  }

  get statusClass() {
    return Event.statusMap[this.status];
  }

  get assigneeLabels() {
    return this.assignee ? this.assignee.join(',') : '--';
  }

  get chartTitle() {
    return this.metric_display[0];
  }
}

